edlib/include/edlib.h

```diff
@@ -7,25 +7,15 @@
  * @brief Main header file, containing all public functions and structures.
  */
 
-// Define EDLIB_API macro that, when on Windows and building a DLL,
-// exports a function/struct so it is visible in the generated DLL.
-// While this happens by default on Unix, it does not
-// on Windows, and DLL will not be successfully linked, so we have to
-// explicitely put this macro in front of every function/struct in here
-// that is part of the Edlib API.
-// EDLIB_DLL_EXPORT should be defined in the .cpp that implements this header.
-// Check discussion here for more details:
-//   https://github.com/Martinsos/edlib/pull/165
-// or check this SO answer:
-//   https://stackoverflow.com/a/538179/1509394
-#if defined(_WIN32) || defined(_WIN64) || defined(__MINGW32__)
-#    ifdef EDLIB_DLL_EXPORT
-#        define EDLIB_API __declspec(dllexport)
-#    else
-#        define EDLIB_API __declspec(dllimport)
-#    endif
+// Define EDLIB_API macro to properly export symbols
+#ifndef EDLIB_SHARED
+#define EDLIB_API
+#elif !defined(_WIN32)
+#define EDLIB_API __attribute__ ((visibility ("default")))
+#elif defined(EDLIB_BUILD)
+#define EDLIB_API __declspec(dllexport)
 #else
-#    define EDLIB_API
+#define EDLIB_API __declspec(dllimport)
```

Comment:
Cool, thanks for fixing this!

I find if-else-if-else chain with so many negations hard to read, how about we refactor it to this:
```c
#ifdef EDLIB_SHARED
#    ifdef (_WIN32)
#        ifdef EDLIB_BUILD
#            EDLIB_API __declspec(dllexport)
#        else
#            EDLIB_API __declspec(dllimport)
#        endif
#    else
#        EDLIB_API __attribute__ ((visibility ("default")))
#    endif
#else
#    EDLIB_API
#endif
```

It does have a couple more lines, but I find it easier to figure out what is happening.

Is (_WIN32) enough to check if we are compiling on windows? What about _WIN64? And what about __MINGW32__ -> that was probably a mistake on my side, because we actually don't want to know if we are compiling on windows, we want to know if we are compiling with msvc? And __mingw32__ should be for mingw and in that case we are back to gcc, right?

Btw., would you mind explaining this to me the whole reasoning a little bit? What I think I am getting is that if EDLIB_SHARED is defined, that means we are building shared library, and therefore we want symbols exported (because by default on windows they are not exported, and we also set it so in CMAKE for unix(gcc) now). Linker needs these symbols so it can link object files / libraries correctly. On windows, we use __declspec statement for this (for msvc). If not on windows, we use __attribute__ ((visibility ("default"))) statement (for gcc). Are those equivalent statements? I read that __declspec is supported in clang, gcc and msvc, but that it is not considered to be "clean C++" because it is not defined by the standard -> so there is a possibility that some other compiler might not support it, and that is why we use it only for Win?

dllexport vs dllimport -> I get that when building a .dll, we want dllexport, because header is included in the implementation file and we want the resulting object to export the symbols. On the other hand, when somebody else is using Edlib and is importing edlib.h and then linking with the .dll, we don't want header to export anything, instead we want it to import those symbols, so dllimport also makes sense to me. But, who sets EDLIB_SHARED in that situation? This is now on them, whoever is using edlib.h and linking with edlib, to define EDLIB_SHARED, right?

Also, how is it that on Windows we have to care about export vs import, while on non-win we just set visibility default and that is it?

Sorry for so many questions, but I find this really interesting and feel I am close to understanding it so I have to use the opportunity :D.

Reply:
That's a lot of questions, and I'm not sure to have sufficient knowledge to answer in depth.

> I find if-else-if-else chain with so many negations hard to read, how about we refactor it to this:
> 
> ```c
> #ifdef EDLIB_SHARED
> #    ifdef (_WIN32)
> #        ifdef EDLIB_BUILD
> #            EDLIB_API __declspec(dllexport)
> #        else
> #            EDLIB_API __declspec(dllimport)
> #        endif
> #    else
> #        EDLIB_API __attribute__ ((visibility ("default")))
> #    endif
> #else
> #    EDLIB_API
> #endif
> ```
> 
> It does have a couple more lines, but I find it easier to figure out what is happening.

Ok, I will modify this part, it's equivalent.

> Is (_WIN32) enough to check if we are compiling on windows? What about _WIN64? And what about **MINGW32** -> that was probably a mistake on my side, because we actually don't want to know if we are compiling on windows, we want to know if we are compiling with msvc? And **mingw32** should be for mingw and in that case we are back to gcc, right?

dll are a windows concept.
If you create a dll (with an import library), MinGW requires the same attributes than msvc (if symbols are hidden, which is the case now).
Therefore you want `__attribute__ ((dllexport))`, but fortunately for gcc `__declspec(dllexport)` is a synonym of `__attribute__ ((dllexport))`: see https://gcc.gnu.org/onlinedocs/gcc/Microsoft-Windows-Function-Attributes.html.
So yes you want to check Windows, not msvc only.

> Btw., would you mind explaining this to me the whole reasoning a little bit? What I think I am getting is that if EDLIB_SHARED is defined, that means we are building shared library, and therefore we want symbols exported (because by default on windows they are not exported, and we also set it so in CMAKE for unix(gcc) now). Linker needs these symbols so it can link object files / libraries correctly. On windows, we use __declspec statement for this (for msvc). If not on windows, we use **attribute** ((visibility ("default"))) statement (for gcc). Are those equivalent statements? I read that __declspec is supported in clang, gcc and msvc, but that it is not considered to be "clean C++" because it is not defined by the standard -> so there is a possibility that some other compiler might not support it, and that is why we use it only for Win?

`(visibility ("default"))` is for ELF, `dllexport` for dll.

For your last question, I can't anwser.

> dllexport vs dllimport -> I get that when building a .dll, we want dllexport, because header is included in the implementation file and we want the resulting object to export the symbols. On the other hand, when somebody else is using Edlib and is importing edlib.h and then linking with the .dll, we don't want header to export anything, instead we want it to import those symbols, so dllimport also makes sense to me. But, who sets EDLIB_SHARED in that situation? This is now on them, whoever is using edlib.h and linking with edlib, to define EDLIB_SHARED, right?

It depends on how edlib is consumed:
- If consumer relies on pkg-config file or CMake config file (or add_subdirectory), he doesn't have to care about this definition because it will be defined (or not if static) for free at consume time in `CFLAGS`.
- If consumer manually adds include directory and flags of edlib, he has to add this definition if he links to shared edlib.

> Also, how is it that on Windows we have to care about export vs import, while on non-win we just set visibility default and that is it?

I don't think it's windows vs non windows, but Visual Studio vs GNU compilers. Visual Studio hides all symbols by default, GCC does not (including gcc@MinGW).
I would say that always hide all symbols is good practice.

Reply:
Awesome, thanks, you answered most of my questions and it certainly clarified things for me! Especially the part with mingw.

Regarding checking if we are on Windows: I am guessing it is common to just check for `_WIN32`? I just did some searching and read that mingw normally also defines `_WIN32` so that certainly sounds good then. What about cygwin, is that something we should worry about?